spec/support/json_matcher.rb

```diff
@@ -0,0 +1,149 @@
+    Rails.logger.error "Cannot match regexp for key #{key}"
+    return false
+  end
+  true
+end
+
+# Checks if the subtree matches the elements across objects
+# object
+# Parameters:
+# object: Hash that contains a subtree in the key
+# compared_object: hash that contains a subtree in the key
+# key: str
+# Returns
+# true if the subtree matches between objects, or false if it doesnt match
+# if there is no subtree it returns nil
+def check_subtree(object, compared_object, key)
+  return unless compared_object[key].is_a?(Hash)
+
+  unless check_objects(object[key], compared_object[key])
+    Rails.logger.error "Cannot match subtree at #{key}"
+    return false
+  end
+  true
+end
+
+# Checks if the arrays matches the elements across objects
+# object
+# Parameters:
+# object: Hash that contains an array in the key
+# compared_object: hash that contains an array in the key
+# key: str
+# Returns
+# true if the array matches all elements between objects, or false if it doesnt match
+# if there is no array it returns nil
+def check_array(object, compared_object, key)
+  return unless compared_object[key].is_a?(Array)
+
+  unless object[key].is_a?(Array)
+    Rails.logger.error "The element #{key} from the original object is not an array"
+    return false
+  end
+  compared_object[key].each_with_index do |_value, pos|
+    unless check_objects(object[key][pos], compared_object[key][pos])
+      Rails.logger.error "Difference when checking position #{pos} of key #{key}"
+      return false
+    end
+  end
+  true
+end
+
+# Checks if the value matches across objects for the same key
+# object
+# Parameters:
+# object: Hash that contains a value in the key
+# compared_object: hash that contains a value in the key
+# key: str
+# Returns
+# true if the value matches between objects, or false if it doesnt match
+def check_value(object, compared_object, key)
+  unless compared_object[key] == object[key]
+    Rails.logger.error "Cannot match subtree at #{key}"
+    return false
+  end
+  true
+end
+
+# Run a pipeline of checks on two objects. If a step in the pipeline
+# returns true or false, it will return that value as the result of the check
+# but if the step returns nil, it will continue running the next step of
+# the pipeline until all of them have been completed, or any of them has returned
+# a true/false value.
+# If none of the steps of the pipeline returns a boolean at the end, it will consider
+# the final result as true.
+# Parameters:
+# object: Hash
+# compared_object: Hash
+# key: string
+# Returns:
+# false if one of the steps of the pipeline returned false; true if one of the steps
+# of the pipeline returned true, or if all the steps returned nil
+def run_checks_pipeline(object, compared_object, key)
+  pipelines = %w[
+    check_regexp check_subtree check_array check_value
+  ]
+  pipelines.each do |method_name|
+    value = send(method_name, object, compared_object, key)
+    return value unless value.nil?
+  end
+  true
```

Comment:
Each method in the pipelines array returns a boolean; true or false. I think it is never nil, which will cause early exit. I don't think we can reach the last line of the run_checks_pipeline method.

Reply:
The first line of the checks have a 'return' without any arguments, which it will be nil when met the condition. nil means that the check does not apply.